### PR TITLE
Trying to fix crash from IKEv2ConnectionAPIImplMac

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/37535

Assuming that this crash happened when block is running but IKEv2ConnectionAPIImplMac is destroyed.
Trying to use weak_ptr to prevent running if IKEv2ConnectionAPIImplMac is already destoyed.

Resolves

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by simonhong** [comment]
@Brandon-T Need you advice here.
I'm trying to use weak ptr to prevent running block. Curious this is right approach. Thanks!

**Comment by Brandon-T** [comment]
> @Brandon-T Need you advice here. I'm trying to use weak ptr to prevent running block. Curious this is right approach. Thanks!

Your code is correct. Alternatively you can capture/move the weak_ptr into a C++ block. This is because Objective-C++ can use both Objective-C lambdas and C++ lambdas:

Your current code:
```c++
auto weak_ptr = weak_factory_.GetWeakPtr();
[vpn_manager loadFromPreferencesWithCompletionHandler:^(NSError* error) {
    if (!weak_ptr) {
        return;
    }

    // Implicitly capturing "this". The below is equivalent to this->SetLastConnectionError(...)
    SetLastConnectionError(....);
}];
```
vs.
```c++
auto weak_ptr = weak_factory_.GetWeakPtr();
[vpn_manager loadFromPreferencesWithCompletionHandler:[weak_ptr = std::move(weak_ptr)](NSError* error) {
    if (!weak_ptr) {
        return;
    }

    // Cannot capture `this` unless we add it to the lambda capture block. So we use `weak_ptr` explicitly.
    weak_ptr->SetLastConnectionError(....);
}];
```
There's no difference other than capture semantics.

TLDR: Your code is correct.

**Comment by Brandon-T** [comment]
Btw, completely unrelated to this PR, I noticed `GetPasswordRefForAccount` and `StorePassword` should not be called on the `main` thread as they read/write to the Keychain which can block the main-thread. Should probably dispatch those to a `global_queue` or `IO::Thread`.

**Review by Brandon-T — APPROVED** [review]

**Comment by simonhong** [comment]
> Also, completely unrelated to this PR, I noticed `GetPasswordRefForAccount` and `StorePassword` are ran on the main-thread. The should not be called on the `main` thread as they read/write (`SecItemCopyMatching, SecItemAdd, SecItemUpdate`) to the Keychain/Secure-Enclave which can block the main-thread/UI thread. Should probably dispatch those to a `global_queue` or `IO::Thread`.
> 
> Any keychain operations should be posted to an I/O thread. iOS and MacOS compilation via Xcode will give warnings for this, but I doubt it shows it in Chromium compilation tools.
> 
> Just something to note :)

Sweet! Filed https://github.com/brave/brave-browser/issues/37561. Thanks! 👍🏼
